distrib-6-08-ocaml-4-xx : To allow compilation with OCAML 4.06 (for Debian, Ubuntu, ...)

GuillaumeBrochu opened the pull request:
These modifications will be required sooner or later, as OCaml 4.06 broke the compilation of Geneweb 6. It will allow to maintain Geneweb 6 in Debian and Ubuntu, while waiting for the new Geneweb 7.

Most of the changes are from commits taken from the master branch (Thanks to M. Bouaziz!).

I also included the fix for absolute paths in Windows and updated the copyright year.

I'm currently testing these changes with my own genealogy. So far, everything is fine.

mbouaziz committed:
Conversion to -safe-string, first steps

mbouaziz committed:
Use Obj.xxx_tag instead of tags computed at runtime

mbouaziz committed:
Do not need to export these functions

fablhx committed:
FIX #8: Date text with space/underscore.

mbouaziz committed:
Removed duplicate function nl

mbouaziz committed:
kprintf is deprectated

mbouaziz committed:
Iovalue.digest returns a string, not a Digest.t

A Digest.t is a 16-char string but Iovalue.digest returns the hexadecimal representation.

mbouaziz committed:
Iovalue.string_code is Digest.to_hex

mbouaziz committed:
Buff: attempt to use ocaml 4.02 generative functors

Seems like the syntax is not supported by camlp5...

mbouaziz committed:
Buff: optimize mstore and gstore

GuillaumeBrochu committed:
Fix Secure.decompose for absolute paths with drive letters in Windows

See https://github.com/geneweb/geneweb/issues/423

To prevent an infinite recursion when there is a drive letter in the path.

GuillaumeBrochu committed:
Really use -safe-string

Adapted from (M. Bouaziz)
f3cddca6fc533aae146a030d09b91b39bf830e46
ee090da46bf881de7412ea1bab076b629ee867d2

GuillaumeBrochu committed:
Update launch commands for MacOS / add GeneWeb.command to distrib

Thanks to Henri Gouraud

GuillaumeBrochu commented:
Henri tested this PR on MacOS (with Ocaml 4.06) and, up to now, it works.

The last commit is aimed to give a better experience to Mac users!

GuillaumeBrochu commented:
Is it possible to merge this one to distrib-6-08-ocaml-4-xx?

fablhx commented:
I'll check that in the coming weekend

fablhx commented:
This PR is very tricky. First I think we should agree on the purpose of this PR. Correct me if I'm wrong, but the purpose is to build the last available version of Geneweb (6.08) for distros (ubuntu, debian and so on).

If we agree on that, I would like very much that the PR only uses the commits that fix the compilation problem and do not add any feature.

fablhx commented:
Here is a more precise comment:

Where does this modification comes from ? I would like to understand this modification
- Minor modification to the previous commit, to fix compilation

This commit should not be added:
- Update copyright year (2018)
- FIX #8: Date text with space/underscore.
- Revert "Warning 3 (Deprecated feature) disabled, warning 52 not fatal"

This commit should not be added unless there are still compilation warning
- Remove -warn-error A

This commit does not build, see your other PR for a full change
- gui.ml : Fix compilation problem with OCAML 4.06.0 (V6)

I wish this commit to be unnecessary, but we can keep it
- Update launch commands for MacOS / add GeneWeb.command to distrib

GuillaumeBrochu commented:
The purpose of this branch was initially to fix ocaml-related bugs for linux-based distros, but some other users here found it useful to have these fixes for windows and mac. There are also some small bug fixes that do not changes the way the software work and that improve user experience or that help package managers (get rid of some patches for instance).

``Minor modification to the previous commit, to fix compilation``
merged with the "previous commit" from M. Bouaziz. Just a minor refactoring. I have included a reference to the original commit.

``Update copyright year (2018)``
removed...

``FIX #8: Date text with space/underscore.``
Required with Ocaml 4.06.0. Otherwise, I get :
``File "gwu.ml", line 117, characters 49-69:
Error: Unbound value spaces_to_underscore``

``Revert "Warning 3 (Deprecated feature) disabled, warning 52 not fatal"``
~~Required to be able to cherry-pick the commits from M. Bouaziz without conflict~~
squashed with ``Really use -safe-string``

``Remove -warn-error A``
~~Required as there are still warnings to be fixed (3 and 52). Since these warnings are not critical yet (for ocaml 4.06.0) and since there is a problem with capitalize_ascii (https://github.com/geneweb/geneweb/issues/593), I think it's preferable to wait to fix them. Current master branch also uses `-warn-error A`.~~
no longer needed

``gui.ml : Fix compilation problem with OCAML 4.06.0 (V6)``
~~Actually, it works with Ocaml 4.06.0, but it looks there is a problem with AppVeyor with ocaml 4.05.0.~~ 
~~This is related [to this change in lablgtk](http://git.ocamlcore.org/cgi-bin/gitweb.cgi?p=lablgtk/lablgtk.git;a=blobdiff;f=src/gMain.mli;h=2a161d1f6dd0259f86afbb1c2a99b325aad6d0e0;hp=3fd8bd4eaa33402d2b0bd5f99dd04d4446d4cfe4;hb=c37e2ef1ad4cc08958f9a49357923de0f21ec93c;hpb=2e9eaac675adc36053a602935fef003d123bd4b6)~~
~~Any ideas to make it work for both ocaml 4.05 AND ocaml 4.06???~~
REMOVED

GuillaumeBrochu commented:
Updated by removing this commit : `gui.ml : Fix compilation problem with OCAML 4.06.0 (V6)`

Ready to be merged into distrib-6-08-ocaml-4-xx, unless there is another problem.

GuillaumeBrochu commented:
Is it possible to review it again?

I've squashed some commits to clarify things.  

No that this commit taken from the master branch:
``FIX #8: Date text with space/underscore.``
is still required with Ocaml 4.06.0, otherwise, I get :
``File "gwu.ml", line 117, characters 49-69:
Error: Unbound value spaces_to_underscore``

GuillaumeBrochu commented:
Bonjour,

Pour votre info., il y a encore plusieurs usagers qui utilisent geneweb sur Debian/Raspbian/Ubuntu. Exemple :
https://www.geneanet.org/forum/viewtopic.php?f=85&t=581107#p1340407

En attendant une version 7 finalisée avec gwc2, cette version 6 est tout ce qui reste d'utilisable (avec gwc2) et de compilable avec un OCAML récent.

Si vous pouviez accepter cette PR (ou me dire ce qu'il faut faire pour la rendre acceptable...), je pourrais continuer de travailler sur le dépôt Debian. Presque tout ce qui s'y trouve vient de commits backportés de la branche master. 

J'utilise personnellement l'équivalent de cette PR pour ma généalogie personnelle (je l'ai compilée avec Ocaml 4.06 en guise de test, mais elle se compile aussi avec Ocaml 4.05). Henri l'a aussi testé de son côté sur Mac­. Tout fonctionne bien jusqu'à preuve du contraire. 

Ocaml 4.06 n'est pas encore arrivé dans Debian, mais dès qu'il arrivera dans Sid, Geneweb sera alors en [FTBFS](https://fr.wikipedia.org/wiki/FTBFS). Merci de m'aider à prévenir ceci :-)

GuillaumeBrochu commented:
Merci
